txpipe/covariance.py

```diff
@@ -61,206 +77,419 @@ def read_cosmology(self):
+        second_is_shear_shear = ('source' in tracer_comb2[0]) and ('source' in tracer_comb2[1])
+
+        if self.do_xi and (first_is_shear_shear or second_is_shear_shear):
+            # this adds the B-mode shape noise contribution.
+            # We assume B-mode power (C_ell) is 0
+            Bmode_F = 1
+            if xi_plus_minus1 != xi_plus_minus2:
+                # in the cross term, this contribution is subtracted.
+                # eq. 29-31 of https://arxiv.org/pdf/0708.0387.pdf
+                Bmode_F=-1 
+            cov[1324] += np.outer(cl[13]*0 + SN[13], cl[24]*0 + SN[24]) * coupling_mat[1324] * Bmode_F
+            cov[1423] += np.outer(cl[14]*0 + SN[14], cl[23]*0 + SN[23]) * coupling_mat[1423] * Bmode_F
+
+        cov['final']=cov[1423]+cov[1324]
+
+        if self.do_xi:
+            s1_s2_1 = self.get_spins(tracer_comb1)
+            s1_s2_2 = self.get_spins(tracer_comb2)
+
+            # For the shear-shear we have two sets of spins, plus and minus,
+            # which are returned as a dict, so we need to pull out the one we need
+            # Otherwise it's just specified as a tuple, e.g. (2,0)
+            if isinstance(s1_s2_1, dict):
+                s1_s2_1 = s1_s2_1[xi_plus_minus1]
+            if isinstance(s1_s2_2, dict):
+                s1_s2_2 = s1_s2_2[xi_plus_minus2]
+
+            # Use these terms to project the covariance from C_ell to xi(theta)
+            th, cov['final']=WT.projected_covariance2(
+                l_cl=ell, s1_s2=s1_s2_1, s1_s2_cross=s1_s2_2, cl_cov=cov['final'])
+
+        # Normalize
+        cov['final'] /= norm
+
+        # Put the covariance into bins. 
+        # This is optional in the case of a C_ell covariance (only if bins in ell are
+        # supplied, otherwise the matrix is for each ell value individually).  It is
+        # required for real-space covariances since these are always binned.
+        if self.do_xi:
+            thb, cov['final_b'] = bin_cov(r=th/d2r, r_bins=ell_bins, cov=cov['final'])
+        else:
+            if ell_bins is not None:
+                lb, cov['final_b'] = bin_cov(r=ell, r_bins=ell_bins, cov=cov['final'])
+        return cov
+    
+    def get_angular_bins(self, two_point_data):
+        # Assume that the ell binning is the same for each of the bins.
+        # This is true in the current pipeline.
+        X = two_point_data.get_data_points('galaxy_shear_cl_ee',i=0,j=0)
+        # Further assume that the ell ranges are contiguous, so that
+        # the max value of one window is the min value of the next.
+        # So we just need the lower edges of each bin and then the
+        # final maximum value of the last bin
+        ell_edges = [x['window'].min for x in X]
+        ell_edges.append(X[-1]['window'].max)
+
+        return np.array(ell_edges)
+
+    #compute all the covariances and then combine them into one single giant matrix
+    def compute_covariance(self, cosmo, meta, two_point_data={}):
```

Comment:
is two_point_data really an optional argument here?

Reply:
ah no, fixed